### PR TITLE
Be stricter about which source paths we load values from

### PR DESCRIPTION
**Pull request by romseygeek**
`XContentMapValues.extractRawValues()` contains a hack whereby we
ignore extra 'dummy' dotted path values appended to a source path - ie,
loading data from `foo.bar.dummy` will also load from `foo.bar`.  This was
used to implement source loading for multifields, which do not have a name
that corresponds to an existing path in source data.  This was safe in general
because you can't have a field that is a dot prefix of another field in index
mappings.

With runtime fields, however, this changes.  You can define scriptless runtime
fields for both `foo.bar` and `foo.bar.dummy`, and it is important that the
latter does not pick up values from the former path.

Source loading for multifield highlighting has already moved away from using
this hack by loading everything through value fetchers.  This commit removes
the hack entirely, and updates term vector loading to use a value fetcher
as well

### CODE REVIEW

**Comment by romseygeek** [comment]
The termvectors trap was caught by a test, but we should review everywhere that is calling `XContentMapValues.extractRawValues()` to make sure that they're not depending on this hack to load multifield values properly.

File: server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java [hunk 1]
```diff
@@ -302,6 +301,16 @@ public void testExtractRawValueMixedObjects() throws IOException {
...
             }
         }
+        {
+            XContentBuilder builder = XContentFactory.jsonBuilder().startObject()
+                .field("foo", "bar")
+                .field("foo.subfoo", "baz")
+                .endObject();
+            try (XContentParser parser = createParser(JsonXContent.jsonXContent, Strings.toString(builder))) {
+                assertThat(XContentMapValues.extractRawValues("foo.subfoo", parser.map()),
+                    containsInAnyOrder("baz"));
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java:311]: were we always returning also foo's values or only when foo.subfoo did not exist?

Comment (romseygeek) [reply server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java:311]: Always both of them.

**Comment by romseygeek** [comment]
Some interesting failures here:
* `380_global` - relies on the standard dynamic parsing which automatically creates a `.keyword` multifield for text, and then aggregates on it.  For runtime fields, this `.keyword` field obviously doesn't exist.  We can either exclude this from the runtime field conversion list or change the test to explicitly map the string field
* `370_doc_count_field` - also refers to a `.keyword` field but it's using a standard mapping so `.keyword` shouldn't actually exist here.  Maybe something weird going on with how terms aggs load their data? I'll investigate, but I think we can actually remove the `.keyword` suffix from the test anyway as it shouldn't be there.
* `300_pipeline` - another one relying on standard dynamic parsing to create a `.keyword` multifield, so the same comments apply as on `380_global`

File: server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java (hunk 1 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java:311]: oh boy, it looks like a bad bug at first glance, on the other hand I am thinking that given that we used to dynamically map objects in the properties section, it is not so easy to hit it, yet good to fix it now that objects are no longer mapped when dynamic is set to runtime which allows for even more free mappings.

**Comment by romseygeek** [comment]
Gonna ask @nik9000 to have a quick look at this as well seeing as he wrote the test converter stuff

**Comment by javanna** [comment]
I am a bit confused about the failures and corresponding tests that need skipping. When we test runtime fields we don't convert tests that rely on the multi-field text and keyword, because we can't support it (we would like to have a text main field and a keyword runtime subfield but there is no way to achieve that dynamically)?

**Comment by javanna** [comment]
Do you have more info on the failures? Besides my comment above, I find it surprising that fixing a bug in how we load values from _source causes us to need to skip more tests.

**Comment by nik9000** [comment]
I haven't read it yet, but IIRC when I first did this work I only skipped when we required the `text` field - I *thought* we could support `.keyword` subfields with runtime fields because, well, we support `keyword` type fields and we can load data from anywhere. I'll read more in a bit though.

**Comment by romseygeek** [comment]
These tests *do* rely on a dynamically mapped text-and-keyword-subfield combination.  I'm not sure why that's not detected by the test converter though.

**Comment by javanna** [comment]
I am doing some digging into this, also trying to remember how we run these qa tests. What I said around skipping tests that rely on dynamically mapped strings is true for the core with mapped tests, which are all green. The failures are all on the qa tests for runtime fields on the search request (core with search). In that case, I suppose that we do create a scriptless runtime field with `.keyword` suffix as part of the runtime_mappings, and while that one used to return the value even if it was in _source in the parent field, it no longer does now that you fixed the bug/leniency?

**Comment by romseygeek** [comment]
> it no longer does now that you fixed the bug/leniency?

That would make sense

**Comment by javanna** [comment]
This fixes it for me: https://github.com/javanna/elasticsearch/commit/28e91390a80a116e8fb6078d2e3928aa6f7b066b . We need to create a runtime field with a small script for this case, that specifically looks for the field one level above.

**Comment by romseygeek** [comment]
Makes sense to me @javanna, I'll merge that change in and back out the blocklisting.  Thanks!

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Comment by romseygeek** [comment]
@javanna I pushed your suggested change, I think this is good to go?

**Review by javanna — APPROVED** [review]
LGTM

**Comment by romseygeek** [comment]
@elasticmachine run elasticsearch-ci/docs

**Pull request merged**